pkg/processor/runtime/python/py/_nuclio_wrapper_as_server.py

```diff
@@ -0,0 +1,308 @@
+import asyncio
+import selectors
+
+from wrapper_common import (
+    WrapperFatalException,
+    AbstractWrapper,
+    create_logger,
+    get_parser_with_common_args)
+
+
+class Wrapper(AbstractWrapper):
+    def __init__(self,
+                 logger,
+                 loop,
+                 handler,
+                 serving_address,
+                 control_socket_path,
+                 platform_kind,
+                 namespace=None,
+                 worker_id=None,
+                 trigger_kind=None,
+                 trigger_name=None,
+                 decode_event_strings=True,
+                 max_connections=None):
+        super().__init__(
+            logger,
+            loop,
+            handler,
+            control_socket_path,
+            platform_kind,
+            namespace,
+            worker_id,
+            trigger_kind,
+            trigger_name,
+            decode_event_strings,
+        )
+        # Validate that the handler is an async function
+        if not asyncio.iscoroutinefunction(self._entrypoint):
+            raise WrapperFatalException(f"The provided handler '{self._entrypoint.__name__}' "
+                                        f"must be an async function (async def).")
+        split_address = serving_address.split(":")
+        self._host = split_address[0]
+        self._port = int(split_address[1])
+
+        # Max file descriptors allowed by the OS
+        self._max_connections = max_connections if max_connections else os.sysconf("SC_OPEN_MAX")
+        self.selector = selectors.DefaultSelector()
+        self.connections = {}  # Track active sockets
+        self.tasks = {}  # Track active asyncio tasks
+        self.shutdown_event = asyncio.Event()
+
+    async def start(self):
+        """Start the server."""
+        # Create server socket
+        server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        server_sock.setsockopt(
+            # Level: Applies the option at the socket level
+            socket.SOL_SOCKET,
+            # Option: Allows reuse of the address/port
+            socket.SO_REUSEADDR,
+            # Value: Enables the option (0 would disable it)
+            1,
+        )
+        server_sock.bind((self._host, self._port))
+        server_sock.listen(self._max_connections)
+        server_sock.setblocking(False)
+
+        self._logger.info(f"Python wrapper server started on {self._host}:{self._port}, "
+                          f"max_connections: {self._max_connections}")
+
+        # Register server socket with default os selector (epoll for Linux)
+        self.selector.register(server_sock, selectors.EVENT_READ)
+        self.connections[server_sock.fileno()] = server_sock
+        try:
+            while True:
+                events = await self.select_events()
+                # Now awaiting the select operation
+                for key, event in events:
+                    sock = key.fileobj
+                    if sock == server_sock:
```

Comment:
In what case will we get a different socket here?
Or is it just being defensive?

Reply:
@TomerShor it would be possible if we subscribed on other events, so this is just defensive